In [ ]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import matplotlib.pyplot as plt
import os
import glob

Check current runtime

In [ ]:
print("Tensorflow version: ",tf.__version__)
print("Available GPUs:", tf.config.list_physical_devices('GPU')) # check for GPU

Tensorflow version:  2.19.0
Available GPUs: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


Download & un the data

In [ ]:
def get_data_extract():
  if "dataset" in os.listdir():
    print("Dataset already exists")
  else:
    print("Downloading the data...")
    !wget -O food-data.zip https://www.kaggle.com/api/v1/datasets/download/trolukovich/food11-image-dataset
    print("Dataset downloaded!")
    print("Extracting data..")
    !mkdir dataset
    !unzip -q food-data.zip -d dataset
    print("Extraction done!")

get_data_extract()

--2026-03-15 02:57:10--  https://www.kaggle.com/api/v1/datasets/download/trolukovich/food11-image-dataset
Resolving www.kaggle.com (www.kaggle.com)... 35.244.233.98
Connecting to www.kaggle.com (www.kaggle.com)|35.244.233.98|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://storage.googleapis.com:443/kaggle-data-sets/432700/821742/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260315%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260315T025710Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=9bcb254e27c23e85851de3072a025999ab4eec470c176ea87476f2f30b977bef5d699ca0868c7bae3c0a8af6c217e1c027afd4a7fd38188f467dccd38d9f457422a9faa9b862bca059349cb68ca21085ad64cada7f798e80ce0f1f0796947361cd7d225fbc93f271287f9eebadc127d39a4fb2c50114039d105f8a6fbed9d624722538a6c82e17ff38f8f1e30867343ebb9eaa0d353e9bb91a93480ecb598b477a256bef042a61ad87428e01bfb435fa103b88eec70122

# Dataset

Explore folders

In [ ]:
for root, dirs, files in os.walk("dataset"):
    print(root)
    if dirs:
        print(" Subfolders:", dirs[:10])

dataset
 Subfolders: ['training', 'validation', 'evaluation']
dataset/training
 Subfolders: ['Egg', 'Soup', 'Noodles-Pasta', 'Rice', 'Seafood', 'Bread', 'Meat', 'Dairy product', 'Vegetable-Fruit', 'Dessert']
dataset/training/Egg
dataset/training/Soup
dataset/training/Noodles-Pasta
dataset/training/Rice
dataset/training/Seafood
dataset/training/Bread
dataset/training/Meat
dataset/training/Dairy product
dataset/training/Vegetable-Fruit
dataset/training/Dessert
dataset/training/Fried food
dataset/validation
 Subfolders: ['Egg', 'Soup', 'Noodles-Pasta', 'Rice', 'Seafood', 'Bread', 'Meat', 'Dairy product', 'Vegetable-Fruit', 'Dessert']
dataset/validation/Egg
dataset/validation/Soup
dataset/validation/Noodles-Pasta
dataset/validation/Rice
dataset/validation/Seafood
dataset/validation/Bread
dataset/validation/Meat
dataset/validation/Dairy product
dataset/validation/Vegetable-Fruit
dataset/validation/Dessert
dataset/validation/Fried food
dataset/evaluation
 Subfolders: ['Egg', 'Soup', 'Noodles

Create Dataset from list of path

In [ ]:
train_path = glob.glob("dataset/training/*/*.jpg")
val_path = glob.glob("dataset/validation/*/*.jpg")
test_path = glob.glob("dataset/evaluation/*/*.jpg")

train_label = [i.split(".")[0].split("/")[-2] for i in train_path]
val_label = [i.split(".")[0].split("/")[-2] for i in val_path]
test_label = [i.split(".")[0].split("/")[-2] for i in test_path]

print("Train images:", len(train_path))
print("Validation images:", len(val_path))
print("Test images:", len(test_path))

Train images: 9866
Validation images: 3430
Test images: 3347


Image augmentation

In [ ]:
from tensorflow.keras.applications.efficientnet import preprocess_input

image_width, image_height = 224, 224

aug = keras.Sequential([
    keras.layers.Resizing(image_width, image_height),
    keras.layers.RandomFlip("horizontal"),
    keras.layers.RandomRotation(0.1),
    keras.layers.RandomZoom(0.1),
    keras.layers.Lambda(preprocess_input),
])

Dataloader

In [ ]:
def load_image(path, label, aug):
    image = tf.io.read_file(path)
    image = tf.image.decode_jpeg(image, channels=3)
    # image augmentation
    image = aug(image)
    return image, label

In [ ]:
class_names = sorted(list(set(train_label)))
class_to_index = {name: idx for idx, name in enumerate(class_names)}

train_label = [class_to_index[i] for i in train_label]
val_label = [class_to_index[i] for i in val_label]
test_label = [class_to_index[i] for i in test_label]

num_classes = len(class_names)

print("Classes:", class_names)
print("Number of classes:", num_classes)

Classes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Number of classes: 11


In [ ]:
train_ds = tf.data.Dataset.from_tensor_slices((train_path, train_label))
train_ds = train_ds.shuffle(buffer_size=len(train_path), reshuffle_each_iteration=False)
train_ds = train_ds.map(lambda x, y: load_image(x, y, aug)).batch(16).prefetch(tf.data.AUTOTUNE)

In [ ]:
val_ds = tf.data.Dataset.from_tensor_slices((val_path, val_label))
val_ds = val_ds.map(lambda x, y: load_image(x, y, aug)).batch(16).prefetch(tf.data.AUTOTUNE)

test_ds = tf.data.Dataset.from_tensor_slices((test_path, test_label))
test_ds = test_ds.map(lambda x, y: load_image(x, y, aug)).batch(16).prefetch(tf.data.AUTOTUNE)

Display Sample from our custom dataset

In [ ]:
images, labels = next(iter(train_ds))
print("Images shape:", images.shape)
print("Labels:", labels.numpy())

Images shape: (16, 224, 224, 3)
Labels: [ 0  9 10  0  5  5 10 10  0  2  1  3  3  4  0  1]


# Model

 Experiment 1: Feature Extraction

In [ ]:


model_version_fe = "EfficientNetB0 - Feature Extraction"

base_model = tf.keras.applications.EfficientNetB0(
    include_top=False,
    weights="imagenet",
    input_shape=(224, 224, 3)
)

# Freeze all base layers
base_model.trainable = False

inputs = keras.Input(shape=(224, 224, 3))
x = base_model(inputs, training=False)
x = keras.layers.GlobalAveragePooling2D()(x)
x = keras.layers.Dropout(0.3)(x)
outputs = keras.layers.Dense(num_classes, activation="softmax")(x)

model = keras.Model(inputs, outputs)

model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

print("Model Version:", model_version_fe)
model.summary()

Model Version: EfficientNetB0 - Feature Extraction


Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer_5 (InputLayer)      │ (None, 224, 224, 3)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ efficientnetb0 (Functional)     │ (None, 7, 7, 1280)     │     4,049,571 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling2d_1      │ (None, 1280)           │             0 │
│ (GlobalAveragePooling2D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 1280)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 11)             │        14,091 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 4,063,662 (15.50 MB)

 Trainable params: 14,091 (55.04 KB)

 Non-trainable params: 4,049,571 (15.45 MB)

In [ ]:
callbacks = [
    keras.callbacks.EarlyStopping(
        monitor="val_loss",
        patience=3,
        restore_best_weights=True
    ),
    keras.callbacks.ReduceLROnPlateau(
        monitor="val_loss",
        factor=0.2,
        patience=2,
        verbose=1
    )
]

In [ ]:
history_fe = model.fit(
    train_ds,
    validation_data=val_ds,
    epochs=10,
    callbacks=callbacks
)

Epoch 1/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 242s 357ms/step - accuracy: 0.7462 - loss: 0.8254 - val_accuracy: 0.8254 - val_loss: 0.5520 - learning_rate: 0.0010
Epoch 2/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 200s 324ms/step - accuracy: 0.8351 - loss: 0.5096 - val_accuracy: 0.8449 - val_loss: 0.4742 - learning_rate: 0.0010
Epoch 3/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 201s 325ms/step - accuracy: 0.8549 - loss: 0.4472 - val_accuracy: 0.8496 - val_loss: 0.4476 - learning_rate: 0.0010
Epoch 4/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 202s 327ms/step - accuracy: 0.8640 - loss: 0.4154 - val_accuracy: 0.8621 - val_loss: 0.4271 - learning_rate: 0.0010
Epoch 5/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 202s 327ms/step - accuracy: 0.8712 - loss: 0.3940 - val_accuracy: 0.8592 - val_loss: 0.4245 - learning_rate: 0.0010
Epoch 6/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 200s 325ms/step - accuracy: 0.8720 - loss: 0.3813 - val_accuracy: 0.8621 - val_loss: 0.4228 - learning_rate: 0.0010
Epoch 7/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 202s 327ms/step - accura

In [ ]:
fe_val_loss, fe_val_acc = model.evaluate(val_ds)
fe_test_loss, fe_test_acc = model.evaluate(test_ds)

print("Feature Extraction Validation Loss:", fe_val_loss)
print("Feature Extraction Validation Accuracy:", fe_val_acc)
print("Feature Extraction Test Loss:", fe_test_loss)
print("Feature Extraction Test Accuracy:", fe_test_acc)

215/215 ━━━━━━━━━━━━━━━━━━━━ 54s 251ms/step - accuracy: 0.8688 - loss: 0.4023
210/210 ━━━━━━━━━━━━━━━━━━━━ 63s 300ms/step - accuracy: 0.8826 - loss: 0.3691
Feature Extraction Validation Loss: 0.4022826850414276
Feature Extraction Validation Accuracy: 0.8688046932220459
Feature Extraction Test Loss: 0.36913931369781494
Feature Extraction Test Accuracy: 0.8825814127922058


Experiment 2: Fine-Tuning

In [38]:
model_version_ft = "EfficientNetB0 - Fine-Tuning (last 10 layers)"

base_model.trainable = True

# Freeze all layers except the last 10 layers
for layer in base_model.layers[:-10]:
    layer.trainable = False

print("Model Version:", model_version_ft)
print("Trainable layers in base model:")

for layer in base_model.layers[-15:]:
    print(layer.name, layer.trainable)

Model Version: EfficientNetB0 - Fine-Tuning (last 10 layers)
Trainable layers in base model:
block7a_expand_bn False
block7a_expand_activation False
block7a_dwconv False
block7a_bn False
block7a_activation False
block7a_se_squeeze True
block7a_se_reshape True
block7a_se_reduce True
block7a_se_expand True
block7a_se_excite True
block7a_project_conv True
block7a_project_bn True
top_conv True
top_bn True
top_activation True


In [39]:
model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-6),
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

In [40]:
ft_callbacks = [
    keras.callbacks.EarlyStopping(
        monitor="val_loss",
        patience=3,
        restore_best_weights=True
    )
]

In [41]:
history_ft = model.fit(
    train_ds,
    validation_data=val_ds,
    epochs=10,
    callbacks=ft_callbacks
)

Epoch 1/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 258s 374ms/step - accuracy: 0.8046 - loss: 0.6227 - val_accuracy: 0.8303 - val_loss: 0.5354
Epoch 2/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 205s 332ms/step - accuracy: 0.7997 - loss: 0.6182 - val_accuracy: 0.8329 - val_loss: 0.5346
Epoch 3/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 203s 329ms/step - accuracy: 0.8081 - loss: 0.6104 - val_accuracy: 0.8332 - val_loss: 0.5311
Epoch 4/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 202s 328ms/step - accuracy: 0.8094 - loss: 0.5900 - val_accuracy: 0.8382 - val_loss: 0.5148
Epoch 5/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 290s 374ms/step - accuracy: 0.8081 - loss: 0.5853 - val_accuracy: 0.8344 - val_loss: 0.5151
Epoch 6/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 203s 329ms/step - accuracy: 0.8178 - loss: 0.5709 - val_accuracy: 0.8359 - val_loss: 0.4981
Epoch 7/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 203s 328ms/step - accuracy: 0.8132 - loss: 0.5730 - val_accuracy: 0.8388 - val_loss: 0.5087
Epoch 8/10
617/617 ━━━━━━━━━━━━━━━━━━━━ 203s 328ms/step - accuracy: 0.8167 -

In [42]:
ft_val_loss, ft_val_acc = model.evaluate(val_ds)
ft_test_loss, ft_test_acc = model.evaluate(test_ds)

print("Fine-Tuning Validation Loss:", ft_val_loss)
print("Fine-Tuning Validation Accuracy:", ft_val_acc)
print("Fine-Tuning Test Loss:", ft_test_loss)
print("Fine-Tuning Test Accuracy:", ft_test_acc)

215/215 ━━━━━━━━━━━━━━━━━━━━ 55s 254ms/step - accuracy: 0.8449 - loss: 0.4894
210/210 ━━━━━━━━━━━━━━━━━━━━ 56s 265ms/step - accuracy: 0.8575 - loss: 0.4509
Fine-Tuning Validation Loss: 0.48938313126564026
Fine-Tuning Validation Accuracy: 0.844897985458374
Fine-Tuning Test Loss: 0.4508574604988098
Fine-Tuning Test Accuracy: 0.8574843406677246


In [7]:
fe_val_acc = 0.8688
fe_test_acc = 0.8826

ft_val_acc = 0.8449
ft_test_acc = 0.8575

print("========== Final Comparison ==========")
print(f"Feature Extraction - Val Accuracy: {fe_val_acc:.4f}, Test Accuracy: {fe_test_acc:.4f}")
print(f"Fine-Tuning       - Val Accuracy: {ft_val_acc:.4f}, Test Accuracy: {ft_test_acc:.4f}")

========== Final Comparison ==========
Feature Extraction - Val Accuracy: 0.8688, Test Accuracy: 0.8826
Fine-Tuning       - Val Accuracy: 0.8449, Test Accuracy: 0.8575


In [8]:
fe_val_acc = 0.8688
fe_test_acc = 0.8826

ft_val_acc = 0.8449
ft_test_acc = 0.8575

print("========== Final Comparison ==========")
print(f"Feature Extraction - Val Accuracy: {fe_val_acc:.4f}, Test Accuracy: {fe_test_acc:.4f}")
print(f"Fine-Tuning       - Val Accuracy: {ft_val_acc:.4f}, Test Accuracy: {ft_test_acc:.4f}")

========== Final Comparison ==========
Feature Extraction - Val Accuracy: 0.8688, Test Accuracy: 0.8826
Fine-Tuning       - Val Accuracy: 0.8449, Test Accuracy: 0.8575


In [9]:
import pandas as pd

results_df = pd.DataFrame({
    "Experiment": ["Feature Extraction", "Fine-Tuning"],
    "Validation Accuracy": [0.8688, 0.8449],
    "Test Accuracy": [0.8826, 0.8575]
})

results_df

,Experiment,Validation Accuracy,Test Accuracy
0,Feature Extraction,0.8688,0.8826
1,Fine-Tuning,0.8449,0.8575


## Analysis

### Feature Extraction
The feature extraction experiment achieved the best overall performance in this project. The EfficientNetB0 base model was fully frozen, and only the classification head was trained. This approach reached around 86.9% validation accuracy and 88.3% test accuracy, showing strong generalization and stable convergence.

### Fine-Tuning
In the fine-tuning experiment, the last layers of the EfficientNetB0 base model were unfrozen and retrained using a smaller learning rate. Although the model learned successfully, the final validation and test performance remained lower than feature extraction.

### Comparison
Feature extraction outperformed fine-tuning in this task. This suggests that the pretrained EfficientNetB0 features were already highly suitable for the Food-11 dataset, while additional fine-tuning did not improve generalization under the chosen settings.

### Generalization
Feature extraction showed better generalization because its validation and test accuracy were both higher and stable. Fine-tuning achieved reasonable performance, but it did not surpass the frozen-base approach.

### Convergence
Both experiments converged successfully. Feature extraction converged faster and more effectively, while fine-tuning improved more slowly and produced weaker final results.

### Overfitting
There was no severe overfitting in either experiment. However, fine-tuning did not provide additional gains and may have slightly reduced the benefit of the pretrained representation.

## Bonus
MLflow was used to record experiment metrics and compare the performance of Feature Extraction and Fine-Tuning.  
DagsHub can be integrated to manage datasets and experiment tracking in a cloud-based repository.

In [12]:
!pip install mlflow -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.6/40.6 kB 3.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 85.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 82.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 74.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 13.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 10.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 197.1/197.1 kB 16.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.0/77.0 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 131.2/131.2 kB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 829.5/829.5 kB 52.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.3/207.3 kB 18.4 MB/s eta 0:00:00


In [13]:
import mlflow

with mlflow.start_run(run_name="EfficientNetB0_Comparison"):
    mlflow.log_param("model_name", "EfficientNetB0")
    mlflow.log_metric("fe_val_accuracy", 0.8688)
    mlflow.log_metric("fe_test_accuracy", 0.8826)
    mlflow.log_metric("ft_val_accuracy", 0.8449)
    mlflow.log_metric("ft_test_accuracy", 0.8575)

print("MLflow logging completed.")

2026/03/15 15:47:59 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/03/15 15:48:00 INFO mlflow.store.db.utils: Updating database tables


MLflow logging completed.
